[loxone] Add RGBW device support

mmattan opened the pull request:
Support added for the Color Picker V2 subcontrol of a Lightning Controller V2.
This allows a RGBW device to be linked to an OpenHab Color Picker.

Fixes #3679

ppieczul commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxControlColorPickerV2.java:
```diff
@@ -0,0 +1,64 @@
+
+    public HSBType getColor() {
+        String hsvColor = getStateTextValue(STATE_COLOR);
+        HSBType color = new HSBType(hsvColor.substring(hsvColor.indexOf("(") + 1, hsvColor.length() - 1));
```
1) What should we do when color state has format: temp(brightness,kelvin)? 
2) For hsv() values, it should check for the correct format of the text, i.e. it starts with hsv and follows by three numbers in parenthesis separated by commas, before it constructs HSBType.
3) It should still catch IllegalArgumentException for any other parsing errors, which will be thrown by HSBType constructor.

ppieczul commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/handler/LoxoneMiniserverHandler.java:
```diff
@@ -249,6 +253,16 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+            if (control instanceof LxControlColorPickerV2) {
+                LxControlColorPickerV2 colorPicker = (LxControlColorPickerV2) control;
+
+                if (command instanceof HSBType) {
```
According to openHAB docs, Color item can take commands: OnOff, IncreaseDecrease, Percent, HSB. We need to handle those, additonally to HSB.

ppieczul commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/handler/LoxoneMiniserverHandler.java:
```diff
@@ -578,17 +597,20 @@ private boolean addChannel(List<Channel> channels, String itemType, ChannelTypeU
         } else if (control instanceof LxControlDimmer) {
             addChannel(channels, "Dimmer", dimmerTypeId, id, label, "Dimmer", tags);
+        } else if (control instanceof LxControlColorPickerV2) {
+            addChannel(channels, "Color", colorPickerTypeId, id, label + " / Color Picker", "Color Picker", tags);
```
Why label+" / Color Picker" when all other channels are just named by the label only? The channel type can be seen in the channel type.

mmattan commented:
Hi @ppieczul , I have updated the code and implemented the requested changes.
For the handling of the temp(brightness,kelvin) format, I have chosen to convert the Kelvin temperature to an RGB color.

ppieczul commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxControlColorPickerV2.java:
```diff
@@ -0,0 +1,294 @@
+     * @param color the loxone color
+     * @return boolean
+     */
+    private boolean isHsvColor(String color) {
```
It is a one liner and used only once, maybe it would make sense to just use it in the body of the calling method. Similar in is temperatureColor.

ppieczul commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxControlColorPickerV2.java:
```diff
@@ -0,0 +1,294 @@
+            } else if (isHsvColor(color) && colorSplit.length == 3) {
+                hsbColor = new HSBType(m.group(1));
+            } else {
+                throw new IllegalArgumentException("Invalid color: " + color);
```
This exception is not caught in the caller of this method. I suggest to return null and to catch this exception from new HSBType in this method and return null too.

ppieczul commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxControlColorPickerV2.java:
```diff
@@ -0,0 +1,294 @@
+        HSBType hsbColor = null;
+
+        Matcher m = Pattern.compile("\\((.*?)\\)").matcher(color);
+        if (m.find()) {
```
Ok, this way of matching the syntax will effectively work, but it looks a bit like "hacking". Specifically for example hsvabc(1,2,3) will get accepted. Also hsvabc(a,b,c) or hsv(,,,) will get accepted too, where the values are not numbers and it will get passed to the Integer.valueOf causing exception. I think for the sake of purity it would be good to match the syntax of the received command against the exact format that is expected, separately to hsv and to temp. For example hsv((\d+),(\d+),(\d+)) and take each group member (this is not exact regular expression that would work, just an idea example).

ppieczul commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxControlColorPickerV2.java:
```diff
@@ -0,0 +1,294 @@
+     * @param temp the Kelvin temperature
+     * @return the red value
+     */
+    private int calculateRed(int temp) {
```
A minor suggestion: maybe it would be possible to extend the HSBType class constructor with this way of constructing color?

mmattan commented:
I have modified the code and changed the code accordingly.

ppieczul commented on addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxControlColorPickerV2.java:
```diff
@@ -0,0 +1,174 @@
+                // we have a temp(brightness,kelvin) pattern
+                hsbColor = TemperatureHSBType.fromBrightnessTemperature(valueMatcher.group(1));
+            }
+        } catch (Exception e) {
```
We expect only illegal argument exception, please specify it, so other exceptions don't get hidden.

mmattan commented:
I updated the readme and things.xml (I was not aware of that last one).
Thanks for reviewing the code and giving some good recommendations.
I will check the OpenHab core project to extend the HSBType class with the temperature parsing and when accepted, I will change the loxone binding accordingly.

ppieczul commented on addons/binding/org.openhab.binding.loxone/ESH-INF/thing/thing-types.xml:
```diff
@@ -186,6 +186,12 @@
 		<label>Loxone Dimmer</label>
 		<description>Loxone's dimmer control</description>
 	</channel-type>
+	
```
This needs to be indented with tabs, not spaces (requirement for xmls). You may use ctrl-shift-F under Eclipse to auto indent.

ppieczul commented on addons/binding/org.openhab.binding.loxone/README.md:
```diff
@@ -118,6 +118,7 @@ Currently supported controls are presented in the table below.
 | Switch                                                    | [Virtual inputs](https://www.loxone.com/enen/kb/virtual-inputs-outputs/) of switch type<br>[Push-button](https://www.loxone.com/enen/kb/push-button/)                                                                                                                                                                     | `Switch`              ...
 | TextState                                                 | [State](https://www.loxone.com/enen/kb/state/)                                                                                                                                                                                                                                                                            | `String`              ...
 | TimedSwitch                                               | [Stairwell light switch](https://www.loxone.com/enen/kb/stairwell-light-switch/) or [Multifunction switch](https://www.loxone.com/enen/kb/multifunction-switch/)                                                                                                                                                          | `Switch` <br> <br> `N ...
+| ColorPickerV2                                               | [RGBW 24v Dimmer Tree](https://www.loxone.com/enen/kb/rgbw-24v-dimmer-tree/)                                                                                                                                                          | `Color`                               | `HSBType` (set the color of the light)<br>`Decimal`/`Percent` ( ...
```
Please move it up the list - the list is sorted alphabetically.

mmattan commented:
Hi @ppieczul I updated according to the recommandations.  Thanks for the close follow-up!

mmattan commented:
This pull request can be closed since it is included in the #2905 PR
